File: pkg/macro/mdl.go
package macro

// Structural MDL — the Reflect slice.
//
// This is the first primitive of a compression-as-goal drive: instead of an
// asserted "symmetry is good" scorer, we measure how many foreground cells a
// grid gets "for free" once you are allowed to describe it as
//
//     one half  +  a Reflect opcode  +  the residual that breaks the symmetry
//
// The signal is compression SAVINGS = L_full - L_reflect, i.e. the number of
// cells the Reflect primitive explains away. Two consequences make this the
// honest, dark-room-proof formulation:
//
//   1. Blank / erased grids save nothing (no foreground to mirror) => savings 0.
//      So the drive never rewards deleting structure to reach a short state.
//   2. Reflect is not privileged as a goal. It is one grammar primitive; a
//      translate or repeat primitive would count its own savings the same way.
//      Symmetry only "matters" in a given game to the extent it compresses that
//      game's grids. When more primitives exist, none is hand-picked — the agent
//      pursues whichever pays the most bits. (With a single primitive, this slice
//      proves the DELTA-L mechanism; the fully emergent, no-scorer claim lands
//      once a second primitive competes with this one.)
//
// Cost model for the slice: 1 unit per foreground cell (color is matched
// exactly, so it is folded into the equality test rather than priced separately).

// SymmetrySavings returns how many foreground cells are explained for free by
// horizontal and by vertical reflection respectively.
//
// Horizontal axis mirrors column c <-> W-1-c; a right-half cell that is
// foreground AND equals its mirror partner is determined by the left half, so it
// costs nothing to store => 1 unit saved. Vertical axis mirrors row r <-> H-1-r
// symmetrically. Center column/row (odd size) maps to itself and is skipped: it
// carries no pair, hence no saving.
func SymmetrySavings(grid [][]int, bg int) (savingsH, savingsV int) {
	h := len(grid)
	if h == 0 {
		return 0, 0
	}

	// Horizontal reflection: for each row, pair column c with its mirror.
	for r := 0; r < h; r++ {
		w := len(grid[r])
		for c := 0; c < w/2; c++ {
			mc := w - 1 - c
			// Count the RIGHT-side cell when it is foreground and matches the
			// mirror of the left, so the left half determines it.
			if grid[r][mc] != bg && grid[r][mc] == grid[r][c] {
				savingsH++
			}
		}
	}

	// Vertical reflection: pair row r with row h-1-r.
	for r := 0; r < h/2; r++ {
		mr := h - 1 - r
		w := len(grid[r])
		if len(grid[mr]) < w {
			w = len(grid[mr])
		}
		for c := 0; c < w; c++ {
			if grid[mr][c] != bg && grid[mr][c] == grid[r][c] {
				savingsV++
			}
		}
	}
	return savingsH, savingsV
}

// SymmetryPreference is the single scalar drive for the Reflect slice: the best
// compression saving available under either reflection axis. Maximizing it is
// equivalent to minimizing structural description length via the Reflect
// primitive. It is 0 on a blank grid and rises only when real foreground
// structure becomes mirror-consistent, so hill-climbing on it pushes toward
// symmetric STRUCTURE, never toward an empty canvas.
func SymmetryPreference(grid [][]int, bg int) int {
	sh, sv := SymmetrySavings(grid, bg)
	if sh > sv {
		return sh
	}
	return sv
}

// --- Primitive #2: Translate / Repeat (periodicity) ---
//
// A grid with a repeating tile can be described as one tile + a Repeat opcode:
// a cell at distance p from an equal earlier cell is "explained" by the period.
// Savings = foreground cells determined by the best period.
//
// Two honest guards distinguish real periodic STRUCTURE from the dark room's
// periodic door:
//   - period p >= 2 only (p == 1 is "every cell equals its neighbour" = uniform
//     fill, not a tile);
//   - the repeating tile must be NON-TRIVIAL (>= 2 distinct values in the first
//     band), so a solid single-colour region — which is trivially periodic at
//     every period — earns ZERO translate savings. Reflection needs contrast to
//     pay off; translation does not, so without this guard a free optimiser would
//     "compress" by flooding the canvas one colour. The guard makes Translate
//     reward patterns, not floods.

func minRowLen(grid [][]int) int {
	w := -1
	for _, row := range grid {
		if w < 0 || len(row) < w {
			w = len(row)
		}
	}
	if w < 0 {
		return 0
	}
	return w
}

// horizTileNonTrivial reports whether columns [0,p) contain >= 2 distinct values.
func horizTileNonTrivial(grid [][]int, p int) bool {
	seen := map[int]struct{}{}
	for _, row := range grid {
		for c := 0; c < p && c < len(row); c++ {
			seen[row[c]] = struct{}{}
			if len(seen) >= 2 {
				return true
			}
		}
	}
	return false
}

// vertTileNonTrivial reports whether rows [0,p) contain >= 2 distinct values.
func vertTileNonTrivial(grid [][]int, p int) bool {
	seen := map[int]struct{}{}
	for r := 0; r < p && r < len(grid); r++ {
		for _, v := range grid[r] {
			seen[v] = struct{}{}
			if len(seen) >= 2 {
				return true
			}
		}
	}
	return false
}

// TranslateSavings returns the foreground cells explained for free by the best
// non-trivial horizontal period and by the best non-trivial vertical period.
func TranslateSavings(grid [][]int, bg int) (savingsH, savingsV int) {
	h := len(grid)
	if h == 0 {
		return 0, 0
	}
	w := minRowLen(grid)

	for p := 2; p < w; p++ {
		if !horizTileNonTrivial(grid, p) {
			continue
		}
		s := 0
		for r := 0; r < h; r++ {
			for c := p; c < len(grid[r]); c++ {
				if grid[r][c] != bg && grid[r][c] == grid[r][c-p] {
					s++
				}
			}
		}
		if s > savingsH {
			savingsH = s
		}
	}

	for p := 2; p < h; p++ {
		if !vertTileNonTrivial(grid, p) {
			continue
		}
		s := 0
		for r := p; r < h; r++ {
			w2 := len(grid[r])
			if len(grid[r-p]) < w2 {
				w2 = len(grid[r-p])
			}
			for c := 0; c < w2; c++ {
				if grid[r][c] != bg && grid[r][c] == grid[r-p][c] {
					s++
				}
			}
		}
		if s > savingsV {
			savingsV = s
		}
	}
	return savingsH, savingsV
}

// TranslatePreference is the best compression saving available under the
// Repeat primitive (either axis). 0 on a blank or single-colour grid.
func TranslatePreference(grid [][]int, bg int) int {
	sh, sv := TranslateSavings(grid, bg)
	if sh > sv {
		return sh
	}
	return sv
}

// --- Emergent primitive selection ---
//
// A Primitive is one grammar operator paired with its compression-savings
// measure. The agent has NO hand-picked goal: for a given grid it scores every
// primitive and pursues the one that saves the most bits. Adding a new primitive
// = appending here; nothing marks any of them as "the" goal. Symmetry, periodicity
// and (later) counting/topology all compete on the single currency of bits saved,
// and which one "matters" is decided per grid by the data.
type Primitive struct {
	Name    string
	Savings func(grid [][]int, bg int) int
}

// Primitives is the current Core-Knowledge grammar: geometry (Reflect, Translate)
// and number (Count). None is privileged as a goal — they compete on bits saved.
var Primitives = []Primitive{
	{Name: "Reflect", Savings: SymmetryPreference},
	{Name: "Translate", Savings: TranslatePreference},
	{Name: "Count", Savings: NumerosityPreference},
	{Name: "Correspondence", Savings: CorrespondencePreference},
}

// BestPrimitive returns the primitive that compresses grid the most, and its
// saving. On a tie the earlier-listed primitive wins; a saving of 0 means no
// primitive found exploitable structure (the grid is incompressible under the
// current grammar — a signal to grow it, not to assert a goal anyway).
func BestPrimitive(grid [][]int, bg int) (Primitive, int) {
	best := Primitives[0]
	bestSave := best.Savings(grid, bg)
	for _, p := range Primitives[1:] {
		if s := p.Savings(grid, bg); s > bestSave {
			best, bestSave = p, s
		}
	}
	return best, bestSave
}

// BestPrimitiveDelta returns the largest single-primitive compression gain
// between before and after: max over primitives p of Savings_p(after) -
// Savings_p(before). This is deliberately NOT DrivePreference(after) -
// DrivePreference(before) — that diffs two possibly-DIFFERENT argmax
// primitives, so a big dominant primitive (e.g. a whole-background Translate)
// can swamp a smaller primitive's real local improvement (e.g. Correspondence
// matching one template region) out of the delta entirely. Crediting the
// best-IMPROVING axis instead of the best-ABSOLUTE axis lets model-free
// reinforcement (cmd/alphaarc-play) notice a click that helps a locally
// dominated primitive even while a bigger primitive's savings stay flat or
// dip — the aggregation fix for the s5i5/tn36 Correspondence wall.
func BestPrimitiveDelta(before, after [][]int, bg int) int {
	best := Primitives[0].Savings(after, bg) - Primitives[0].Savings(before, bg)
	for _, p := range Primitives[1:] {
		if d := p.Savings(after, bg) - p.Savings(before, bg); d > best {
			best = d
		}
	}
	return best
}

File: python/alphaarc/mdl.py
"""Structural MDL primitives -- Python port of pkg/macro/mdl.go + mdl_number.go.

This is the compression-drive core, ported to Python because the ARC Prize 2026
/ ARC-AGI-3 Kaggle submission is Python-only and runs OFFLINE (no calling our Go
binary). The logic mirrors the Go source 1:1 -- same "bits saved by a regularity"
currency, same degenerate-cheat guards -- so the offline Go tests and these
tests must agree.

A grid is a list[list[int]] (values 0..15); bg is the background colour.
"""

from __future__ import annotations

from typing import Callable, List, Tuple


Grid = List[List[int]]


# --- Primitive #1: Reflect (symmetry) ---

def symmetry_savings(grid: Grid, bg: int) -> Tuple[int, int]:
    """Foreground cells explained for free by horizontal and by vertical reflection.

    Horizontal mirrors column c <-> W-1-c; a right-half foreground cell equal to
    its mirror is determined by the left half => 1 saved. Vertical is the row
    analogue. Center column/row (odd size) maps to itself and is skipped.
    """
    h = len(grid)
    if h == 0:
        return 0, 0
    savings_h = 0
    for r in range(h):
        w = len(grid[r])
        for c in range(w // 2):
            mc = w - 1 - c
            if grid[r][mc] != bg and grid[r][mc] == grid[r][c]:
                savings_h += 1
    savings_v = 0
    for r in range(h // 2):
        mr = h - 1 - r
        w = min(len(grid[r]), len(grid[mr]))
        for c in range(w):
            if grid[mr][c] != bg and grid[mr][c] == grid[r][c]:
                savings_v += 1
    return savings_h, savings_v


def symmetry_preference(grid: Grid, bg: int) -> int:
    sh, sv = symmetry_savings(grid, bg)
    return sh if sh > sv else sv


# --- Primitive #2: Translate / Repeat (periodicity) ---

def _min_row_len(grid: Grid) -> int:
    w = -1
    for row in grid:
        if w < 0 or len(row) < w:
            w = len(row)
    return 0 if w < 0 else w


def _horiz_tile_nontrivial(grid: Grid, p: int) -> bool:
    """True iff columns [0,p) contain >= 2 distinct values (a pattern, not a flood)."""
    seen = set()
    for row in grid:
        for c in range(min(p, len(row))):
            seen.add(row[c])
            if len(seen) >= 2:
                return True
    return False


def _vert_tile_nontrivial(grid: Grid, p: int) -> bool:
    seen = set()
    for r in range(min(p, len(grid))):
        for v in grid[r]:
            seen.add(v)
            if len(seen) >= 2:
                return True
    return False


def translate_savings(grid: Grid, bg: int) -> Tuple[int, int]:
    """Foreground cells explained by the best non-trivial horizontal / vertical period."""
    h = len(grid)
    if h == 0:
        return 0, 0
    w = _min_row_len(grid)
    savings_h = 0
    for p in range(2, w):
        if not _horiz_tile_nontrivial(grid, p):
            continue
        s = 0
        for r in range(h):
            for c in range(p, len(grid[r])):
                if grid[r][c] != bg and grid[r][c] == grid[r][c - p]:
                    s += 1
        if s > savings_h:
            savings_h = s
    savings_v = 0
    for p in range(2, h):
        if not _vert_tile_nontrivial(grid, p):
            continue
        s = 0
        for r in range(p, h):
            w2 = min(len(grid[r]), len(grid[r - p]))
            for c in range(w2):
                if grid[r][c] != bg and grid[r][c] == grid[r - p][c]:
                    s += 1
        if s > savings_v:
            savings_v = s
    return savings_h, savings_v


def translate_preference(grid: Grid, bg: int) -> int:
    sh, sv = translate_savings(grid, bg)
    return sh if sh > sv else sv


# --- Primitive #3: Number / Numerosity (count) ---

def _shape_key(color: int, cells: List[Tuple[int, int]]) -> str:
    """Position-independent signature: colour + sorted offsets from the top-left."""
    minr = min(c[0] for c in cells)
    minc = min(c[1] for c in cells)
    offs = sorted((r - minr, c - minc) for r, c in cells)
    return "c%d:" % color + "".join("%d,%d;" % (r, c) for r, c in offs)


def _components(grid: Grid, bg: int) -> List[Tuple[int, List[Tuple[int, int]]]]:
    """4-connected same-colour foreground components as (colour, cells)."""
    h = len(grid)
    if h == 0:
        return []
    visited = [[False] * len(grid[r]) for r in range(h)]
    comps: List[Tuple[int, List[Tuple[int, int]]]] = []
    for r in range(h):
        for c in range(len(grid[r])):
            if grid[r][c] == bg or visited[r][c]:
                continue
            color = grid[r][c]
            cells: List[Tuple[int, int]] = []
            stack = [(r, c)]
            visited[r][c] = True
            while stack:
                cr, cc = stack.pop()
                cells.append((cr, cc))
                for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                    nr, nc = cr + dr, cc + dc
                    if nr < 0 or nr >= h or nc < 0 or nc >= len(grid[nr]):
                        continue
                    if visited[nr][nc] or grid[nr][nc] != color:
                        continue
                    visited[nr][nc] = True
                    stack.append((nr, nc))
            comps.append((color, cells))
    return comps


def numerosity_savings(grid: Grid, bg: int) -> int:
    """Sum over identical-shape groups of (count-1) * shape_size.

    Guard: single-cell shapes (size < 2) don't count -- scattering pixels is not
    counting objects.
    """
    group_count = {}
    group_cells = {}
    for color, cells in _components(grid, bg):
        key = _shape_key(color, cells)
        group_count[key] = group_count.get(key, 0) + 1
        group_cells[key] = len(cells)
    savings = 0
    for key, cnt in group_count.items():
        s = group_cells[key]
        if s < 2:
            continue
        if cnt >= 2:
            savings += (cnt - 1) * s
    return savings


def numerosity_preference(grid: Grid, bg: int) -> int:
    return numerosity_savings(grid, bg)


# --- Emergent primitive selection ---

class Primitive:
    __slots__ = ("name", "savings")

    def __init__(self, name: str, savings: Callable[[Grid, int], int]):
        self.name = name
        self.savings = savings


# Correspondence is registered by correspondence.py at import time (appended
# there) so this module stays free of the heavier relational primitive.
PRIMITIVES: List[Primitive] = [
    Primitive("Reflect", symmetry_preference),
    Primitive("Translate", translate_preference),
    Primitive("Count", numerosity_preference),
]


def best_primitive(grid: Grid, bg: int) -> Tuple[Primitive, int]:
    """The primitive that compresses grid the most, and its saving (ties: earlier wins)."""
    best = PRIMITIVES[0]
    best_save = best.savings(grid, bg)
    for p in PRIMITIVES[1:]:
        s = p.savings(grid, bg)
        if s > best_save:
            best, best_save = p, s
    return best, best_save


def drive_preference(grid: Grid, bg: int) -> int:
    _, s = best_primitive(grid, bg)
    return s


def best_primitive_delta(before: Grid, after: Grid, bg: int) -> int:
    """Signed delta of whichever primitive moved MOST by absolute value.

    Mirrors the Go fix (commits 606f770 + 4b2f6f3): NOT drive_preference diff
    (argmax-of-argmax hides a small primitive's real local gain behind a big
    dominant one), and NOT plain max(delta) (an inapplicable primitive flat at 0
    then outranks a genuine negative delta and erases punishment). Biggest-mover,
    sign preserved, does both jobs.
    """
    best = 0
    best_abs = -1
    for p in PRIMITIVES:
        d = p.savings(after, bg) - p.savings(before, bg)
        a = d if d >= 0 else -d
        if a > best_abs:
            best, best_abs = d, a
    return best
